### PR TITLE
MySQL Client: Unix Domain Socket support

### LINKED ISSUE #521: MySQL Unix Domain Socket Support
Motivation:
We can connect to the MySQL Server via Unix Domain Socket, see https://dev.mysql.com/doc/refman/8.0/en/server-options.html#option_mysqld_socket

### PR DESCRIPTION
**Pull request by tsegismont**
Resolves #521

### CODE REVIEW

**Comment by tsegismont** [comment]
@BillyYccc @vietj I have tested locally with a mounted volume for the MySQL socket. I can update the MySQL container rule to do the same.
Is it a blocker for you @vietj ?

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/MySQLConnectOptions.java [hunk 1]
```diff
@@ -544,4 +547,16 @@ public JsonObject toJson() {
...
+  @GenIgnore
+  @Override
+  public SocketAddress getSocketAddress() {
+    String path = getProperties().get("socket");
+    return path != null ? SocketAddress.domainSocketAddress(path) : super.getSocketAddress();
+  }
+
+  @GenIgnore
+  public boolean isUsingDomainSocket() {
+    return this.getProperties().containsKey("socket");
```
Comment (BillyYccc) [inline vertx-mysql-client/src/main/java/io/vertx/mysqlclient/MySQLConnectOptions.java:560]: Are we going this way which is different from the usage of postgres client(using host+port)?
The `socket` is MySQL specific property, how about keep representing the unix socket path with `host` field but add additional support for automatically converting `socket` to `host` in URL parser?

File: vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      client.close();
+    }
+  }
+
+  @Test
+  public void uriTest(TestContext context) throws UnsupportedEncodingException {
+    String path = URLEncoder.encode(options.getProperties().get("socket"), "UTF-8");
+    String uri = "mysql://root:password@localhost?socket=" + path;
+    client = MySQLPool.pool(uri);
+    client.getConnection(context.asyncAssertSuccess(pgConnection -> pgConnection.close()));
```
Comment (BillyYccc) [inline vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java:63]: `pgConnection` should be renamed in this class

File: vertx-mysql-client/src/main/java/io/vertx/mysqlclient/MySQLConnectOptions.java (hunk 1 above)
Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/MySQLConnectOptions.java:560]: `socket` is indeed MySQL specific. It is described in our documentation as supported, even if it actually isn't. Perhaps we should support both

Comment (tsegismont) [reply vertx-mysql-client/src/main/java/io/vertx/mysqlclient/MySQLConnectOptions.java]: Done

File: vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java (hunk 2 above)
Comment (tsegismont) [reply vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java]: Done

File: vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import io.vertx.core.VertxOptions;
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import io.vertx.sqlclient.PoolOptions;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.testcontainers.shaded.org.apache.commons.lang.SystemUtils;
```
Comment (BillyYccc) [inline vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java:23]: can we get rid of such shaded dependency of TestContainers?

File: vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+public class MySQLUnixDomainSocketTest {
+
+  private static final String unixSocketFile = System.getProperty("unix.socket.file");
+
+  private MySQLPool client;
+  private MySQLConnectOptions options;
+
+  @Before
+  public void before() {
+    assumeTrue(SystemUtils.IS_OS_UNIX);
```
Comment (BillyYccc) [inline vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java:41]: Unix is ambiguous here and we also haven't determined how we test in OS X environment with Docker.

File: vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java (hunk 3 above)
Comment (tsegismont) [reply vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java]: Done

File: vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java (hunk 4 above)
Comment (tsegismont) [reply vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLUnixDomainSocketTest.java]: Yes, I've updated this check Linux only for now

**Comment by tsegismont** [comment]
Thank you @BillyYccc !

@vietj anything you want to add? If you're fine with this PR, I'd like to merge it asap and backport to 3.9.

**Comment by vietj** [comment]
I'll have a look soon

File: vertx-mysql-client/src/main/asciidoc/index.adoc [hunk 5]
```diff
@@ -91,6 +92,45 @@ You can easily get one from the pool:
 
 Once you are done with the connection you must close it to release it to the pool, so it can be reused.
 
+=== Unix Domain Socket
+
+Sometimes for simplicity, security or performance reasons, it is required to connect via a Unix Domain Socket.
```
Comment (vietj) [inline vertx-mysql-client/src/main/asciidoc/index.adoc:97]: can you add a link to how a section of MySQL doc about domain sockets ?

Comment (tsegismont) [reply vertx-mysql-client/src/main/asciidoc/index.adoc:97]: Sorry I'm not sure what you mean.

Comment (tsegismont) [reply vertx-mysql-client/src/main/asciidoc/index.adoc:97]: Done

**Pull request merged**